src/components/EarnDashboard/components/ProviderDetails/StakingOpportunitiesByPosition.tsx

```diff
@@ -0,0 +1,245 @@
+import { ArrowForwardIcon } from '@chakra-ui/icons'
+import { Button, Flex } from '@chakra-ui/react'
+import { Tag } from '@chakra-ui/tag'
+import type { Asset } from '@shapeshiftoss/asset-service'
+import type { AssetId } from '@shapeshiftoss/caip'
+import { fromAssetId } from '@shapeshiftoss/caip'
+import { bnOrZero } from '@shapeshiftoss/investor-foxy'
+import type { MarketData } from '@shapeshiftoss/types'
+import { DefiAction, DefiProvider } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import qs from 'qs'
+import { useCallback, useMemo } from 'react'
+import { useTranslate } from 'react-polyglot'
+import { useHistory, useLocation } from 'react-router'
+import type { Column, Row } from 'react-table'
+import { Amount } from 'components/Amount/Amount'
+import { ReactTable } from 'components/ReactTable/ReactTable'
+import { AssetCell } from 'components/StakingVaults/Cells'
+import { RawText } from 'components/Text'
+import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
+import { WalletActions } from 'context/WalletProvider/actions'
+import { useWallet } from 'hooks/useWallet/useWallet'
+import { bn } from 'lib/bignumber/bignumber'
+import { trackOpportunityEvent } from 'lib/mixpanel/helpers'
+import { MixPanelEvents } from 'lib/mixpanel/types'
+import type {
+  OpportunityId,
+  StakingEarnOpportunityType,
+} from 'state/slices/opportunitiesSlice/types'
+import {
+  selectAggregatedEarnUserStakingOpportunitiesIncludeEmpty,
+  selectAssets,
+  selectCryptoMarketData,
+} from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+type StakingPositionsByPositionProps = {
+  ids: OpportunityId[]
+}
+
+export type RowProps = Row<StakingEarnOpportunityType>
+
+type CalculateRewardFiatAmountArgs = {
+  assets: Partial<Record<AssetId, Asset>>
+  marketData: Partial<Record<AssetId, MarketData>>
+} & Pick<StakingEarnOpportunityType, 'rewardAssetIds' | 'rewardsAmountsCryptoBaseUnit'>
+
+type CalculateRewardFiatAmount = (args: CalculateRewardFiatAmountArgs) => number
+
+const calculateRewardFiatAmount: CalculateRewardFiatAmount = ({
+  rewardsAmountsCryptoBaseUnit,
+  rewardAssetIds,
+  assets,
+  marketData,
+}) => {
+  if (!rewardAssetIds) return 0
```

Comment:
`AssetIdsTuple` can be 0-length but never undefined
```suggestion
  if (!rewardAssetIds.length) return 0
```

Reply:
Would work without the early return as well but tomato tomato, both make sense